### PR TITLE
Job Declaration Documentation

### LINKED ISSUE #1204: Need Rust docs on `protocols::v2::subprotocols::job-declaration`
# Background
Currently, the `protocols::v2::subprotocols::job-declaration` has already been documented, but it's worth a look to see where to improve it. 

> Note: While this issue addresses the need for doc comments, it will also serve as a means to identify code debt that needs to be cleaned up and addressed in future issues.

# Problem
When a user navigates to this crate crates.io, there is no useful information for the user. This impedes understanding and adoption.

# Solution
Navigate through each crate in `protocols::v2::subprotocols::job-declaration` and add:
- Rust Docs comments (following the formatting structure in [this on-going discussion](https://github.com/stratum-mining/stratum/discussions/970)), to be displayed in `docs.rs`
- a descriptive `README.md` file to be displayed in `crates.io`

While this work is being performed, specific code debt should be looked for and recorded to turn into future issues

### PR DESCRIPTION
**Pull request by jbesraa**
resolves #1204

### CODE REVIEW

File: protocols/v2/subprotocols/job-declaration/src/declare_mining_job.rs [hunk 1]
```diff
@@ -6,45 +6,88 @@ use binary_sv2::{Deserialize, Seq064K, Serialize, ShortTxId, Str0255, B0255, B06
...
+/// A response sent by JDServer acknowledging [`DeclareMiningJob`] message.
 #[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
 #[repr(C)]
 pub struct DeclareMiningJobSuccess<'decoder> {
+    /// A unique identifier for this request.
+    ///
+    /// Must be the same as the received [`DeclareMiningJob::request_id`].
     pub request_id: u32,
+    /// This **may** be the same token as [DeclareMiningJob::mining_job_token] if the pool allows
+    /// to start mining on a non declared job. If the token is different (irrespective of if the
```
Comment (GitGab19) [inline protocols/v2/subprotocols/job-declaration/src/declare_mining_job.rs:66]:
```suggestion
    /// to start mining on a yet-non-declared job. If the token is different (irrespective of if the
```

File: protocols/v2/subprotocols/job-declaration/src/declare_mining_job.rs [hunk 2]
```diff
@@ -6,45 +6,88 @@ use binary_sv2::{Deserialize, Seq064K, Serialize, ShortTxId, Str0255, B0255, B06
...
+    /// downstream is already mining using it), the downstream **must** send a `SetCustomMiningJob`
+    /// message on each connection which wishes to mine using the declared job.
     #[cfg_attr(feature = "with_serde", serde(borrow))]
     pub new_mining_job_token: B0255<'decoder>,
 }
 
-/// ## DeclareMiningJobError (Server -> Client)
+/// A response sent by JDServer rejecting [`DeclareMiningJob`] message.
```
Comment (GitGab19) [inline protocols/v2/subprotocols/job-declaration/src/declare_mining_job.rs:73]:
```suggestion
/// A response sent by JDS rejecting [`DeclareMiningJob`] message.
```

File: protocols/v2/subprotocols/job-declaration/src/declare_mining_job.rs [hunk 3]
```diff
@@ -6,45 +6,88 @@ use binary_sv2::{Deserialize, Seq064K, Serialize, ShortTxId, Str0255, B0255, B06
...
+///
+/// Downstream should consider this as a trigger to fallback into some other Pool/JDServer or solo
```
Comment (GitGab19) [inline protocols/v2/subprotocols/job-declaration/src/declare_mining_job.rs:75]:
```suggestion
/// Downstream should consider this as a trigger to fallback into some other Pool/JDS or do Solo
```

File: protocols/v2/subprotocols/job-declaration/src/identify_transactions.rs [hunk 4]
```diff
@@ -6,18 +6,29 @@ use binary_sv2::{Deserialize, Seq064K, Serialize, U256};
 #[cfg(not(feature = "with_serde"))]
 use core::convert::TryInto;
 
-/// TODO: comment
+/// Message used by JDServer as a response to a [`crate::DeclareMiningJob`] message indicating it
```
Comment (GitGab19) [inline protocols/v2/subprotocols/job-declaration/src/identify_transactions.rs:9]:
```suggestion
/// Message used by JDS as a response to a [`crate::DeclareMiningJob`] message indicating it
```

File: protocols/v2/subprotocols/job-declaration/src/identify_transactions.rs [hunk 5]
```diff
@@ -6,18 +6,29 @@ use binary_sv2::{Deserialize, Seq064K, Serialize, U256};
...
 #[repr(C)]
 pub struct IdentifyTransactions {
+    /// Mining job unique identifier.
+    ///
+    /// This **must** be the same as the received [`crate::DeclareMiningJob::request_id`].
     pub request_id: u32,
 }
 
-/// TODO: comment
+/// Sent by the JDClient in response to an [`IdentifyTransactions`] message to provide the full set
```
Comment (GitGab19) [inline protocols/v2/subprotocols/job-declaration/src/identify_transactions.rs:21]:
```suggestion
/// Message sent by the JDC in response to an [`IdentifyTransactions`] message to provide the full set
```

File: protocols/v2/subprotocols/job-declaration/src/lib.rs [hunk 6]
```diff
@@ -2,10 +2,16 @@
 
 //! # Job Declaration Protocol
 //!
-//! This protocol runs between the Job Declarator and Pool and can be
-//! provided as a trusted 3rd party service for mining farms.
+//! The following crate defines the set of messages used in the Job Declaration Stratum V2
+//! subprotcol.
```
Comment (GitGab19) [inline protocols/v2/subprotocols/job-declaration/src/lib.rs:6]:
```suggestion
//! This crate defines the set of messages used in the Job Declaration subprotcol of Stratum V2.
```

File: protocols/v2/subprotocols/job-declaration/src/provide_missing_transactions.rs [hunk 7]
```diff
@@ -6,29 +6,39 @@ use binary_sv2::{Deserialize, Seq064K, Serialize, B016M};
...
 
-// In order to do block propagation, the JDserver must know all the transactions within the current
-// block template. These transactions are provided by the JDclient to the JDserver as a sequence
-// of short hashes (in the message DeclareMiningJob). The JDserver has a mempool, which uses to
-// identify the transactions from this list. If there is some transaction that it is not in the
-// JDserver memppol, the JDserver sends this message to ask for them. They are specified by their
-// position in the original DeclareMiningJob message, 0-indexed not including the coinbase
-// transaction transaction.
-
+/// A message used by the JDserver to ask for transactions that it did not recognize from
```
Comment (GitGab19) [inline protocols/v2/subprotocols/job-declaration/src/provide_missing_transactions.rs:9]:
```suggestion
/// A message used by the JDS to ask for transactions that it did not recognize from
```

File: protocols/v2/subprotocols/job-declaration/src/provide_missing_transactions.rs [hunk 8]
```diff
@@ -6,29 +6,39 @@ use binary_sv2::{Deserialize, Seq064K, Serialize, B016M};
...
+/// [`crate::DeclareMiningJob`] message.
+///
+/// In order to do block propagation, JDServer must know all the transactions within the current
+/// block template. These transactions are provided by the JDClient to the JDserver as a sequence
+/// of short hashes in the [`crate::DeclareMiningJob::tx_short_hash_list`] message. If JDserver is
```
Comment (GitGab19) [inline protocols/v2/subprotocols/job-declaration/src/provide_missing_transactions.rs:14]:
```suggestion
/// In order to do block propagation, JDS must know all the transactions within the current
/// block template. These transactions are provided by the JDC to the JDS as a sequence
/// of short hashes in the [`crate::DeclareMiningJob::tx_short_hash_list`] message. If JDS is
```

File: protocols/v2/subprotocols/job-declaration/src/provide_missing_transactions.rs [hunk 9]
```diff
@@ -6,29 +6,39 @@ use binary_sv2::{Deserialize, Seq064K, Serialize, B016M};
...
+/// for them. They are specified by their position in the original DeclareMiningJob message,
+/// 0-indexed not including the coinbase transaction transaction.
 #[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
 #[repr(C)]
 pub struct ProvideMissingTransactions<'decoder> {
+    /// Unique Identifier.
+    ///
+    /// Must be the same as the received [`crate::DeclareMiningJob::request_id`].
     pub request_id: u32,
+    /// A list of unrecognized transactions that need to be supplied by the JDClient in full.  They
```
Comment (GitGab19) [inline protocols/v2/subprotocols/job-declaration/src/provide_missing_transactions.rs:25]:
```suggestion
    /// A list of unrecognized transactions that need to be supplied by the JDC in full. They
```

File: protocols/v2/subprotocols/job-declaration/src/submit_solution.rs [hunk 10]
```diff
@@ -6,17 +6,28 @@ use binary_sv2::{Deserialize, Serialize, B032, U256};
 #[cfg(not(feature = "with_serde"))]
 use core::convert::TryInto;
 
-/// TODO: comment
+/// A message sent by JDClient as soon as it finds a valid block.
+///
+/// Upon receiving this message, JDServer should propogate the new block as soon as possible.
+///
+/// Note that JDClient is also expected to share the new block data through `SubmitSolution`
```
Comment (GitGab19) [inline protocols/v2/subprotocols/job-declaration/src/submit_solution.rs:13]:
```suggestion
/// A message sent by JDC as soon as it finds a valid block.
///
/// Upon receiving this message, JDS should propagate the new block as soon as possible.
///
/// Note that JDC is also expected to share the new block data through `SubmitSolution`
```

File: protocols/v2/subprotocols/job-declaration/src/allocate_mining_job_token.rs [hunk 11]
```diff
@@ -6,34 +6,49 @@ use binary_sv2::{Deserialize, Serialize, Str0255, B0255, B064K};
...
+/// A Message used by a JDServer in response to a [`AllocateMiningJobToken`] message.
 #[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
 #[repr(C)]
 pub struct AllocateMiningJobTokenSuccess<'decoder> {
+    /// A unique identifier for pairing the response/request.
+    ///
+    /// This **must** be the same as the received [`AllocateMiningJobToken::request_id`].
     pub request_id: u32,
+    /// A Token that makes the JD Client eligible for committing a mining job for
+    /// approval/transaction declaration or for identifying custom mining job on mining connection.
```
[the remark this answers is not in the document]
Comment (GitGab19) [reply protocols/v2/subprotocols/job-declaration/src/allocate_mining_job_token.rs:31]: Have you missed this one? We always referenced to it with `JDC`, I've never seen `JD Client`

File: protocols/v2/subprotocols/job-declaration/src/provide_missing_transactions.rs [hunk 12]
```diff
@@ -6,29 +6,39 @@ use binary_sv2::{Deserialize, Seq064K, Serialize, B016M};
...
 
-// In order to do block propagation, the JDserver must know all the transactions within the current
-// block template. These transactions are provided by the JDclient to the JDserver as a sequence
-// of short hashes (in the message DeclareMiningJob). The JDserver has a mempool, which uses to
-// identify the transactions from this list. If there is some transaction that it is not in the
-// JDserver memppol, the JDserver sends this message to ask for them. They are specified by their
-// position in the original DeclareMiningJob message, 0-indexed not including the coinbase
-// transaction transaction.
-
+/// Message used by the JDserver to ask for transactions that it did not recognize from
```
Comment (GitGab19) [inline protocols/v2/subprotocols/job-declaration/src/provide_missing_transactions.rs:9]:
```suggestion
/// Message used by JDS to ask for transactions that it did not recognize from
```

**Review by GitGab19 — COMMENTED** [review]
LGTM. Left a couple of nits

**Review by GitGab19 — APPROVED** [review]

**Comment by jbesraa** [comment]
Rebased without further changes

**Pull request merged**